director/projects/api/views/nodes.py

```diff
@@ -0,0 +1,222 @@
+from typing import Optional
+import hashlib
+import json
+
+from django.shortcuts import get_object_or_404
+from drf_yasg.utils import swagger_auto_schema
+from rest_framework import mixins, permissions, serializers, status, viewsets, renderers
+from rest_framework.exceptions import NotAuthenticated, PermissionDenied
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from lib.converter_facade import (
+    ConverterFacade,
+    ConverterIo,
+    ConverterIoType,
+    ConverterContext,
+)
+from lib.conversion_types import ConversionFormatId
+from projects.models import Node
+from projects.views.mixins import ProjectPermissionsMixin, ProjectPermissionType
+
+
+class NodesCreateRequest(serializers.ModelSerializer):
+    """The request data when creating a new node."""
+
+    node = serializers.JSONField(required=True, help_text="The node itself.")
+
+    class Meta:
+        model = Node
+        fields = ["project", "app", "host", "node"]
+        ref_name = None
+
+
+class NodesCreateResponse(serializers.ModelSerializer):
+    """The response data when creating a new node."""
+
+    url = serializers.SerializerMethodField()
+
+    def get_url(self, obj):
+        request = self.context.get("request")
+        return request.build_absolute_uri(obj.get_absolute_url())
+
+    class Meta:
+        model = Node
+        fields = ["key", "url"]
+        ref_name = None
+
+
+class NodeSerializer(NodesCreateResponse):
+    """The response data when retrieving a node."""
+
+    node = serializers.JSONField(source="json", help_text="The node itself.")
+
+    class Meta:
+        model = Node
+        fields = ["project", "app", "host", "key", "url", "node"]
+        ref_name = None
+
+
+class NodesViewSet(
+    mixins.CreateModelMixin,
+    mixins.RetrieveModelMixin,
+    viewsets.GenericViewSet,
+    ProjectPermissionsMixin,
+):
+
+    lookup_url_kwarg = "key"
+    renderer_classes = [renderers.JSONRenderer, renderers.TemplateHTMLRenderer]
+
+    def get_permissions(self):
+        """
+        Get the list of permissions that the current action requires.
+
+        - `create` (i.e. POST) required authentication to prevent unauthenticated
+          users using this view set as a key/value store
+        - `retrieve` (i.e. GET) does not require authentication (although content
+          will be limited in that case)
+        """
+        if self.action == "create":
+            permission_classes = [permissions.IsAuthenticated]
+        else:
+            permission_classes = []
+        return [permission() for permission in permission_classes]
```

Comment:
Does `permission` need to be called here? I think returning the classes is fine?

Reply:
Yes, it does. Confirmed with the docs https://www.django-rest-framework.org/api-guide/viewsets/ and tests fail if just returning classes.